File: backend/tests/factories/micro_school.py
"""Micro-school factories."""

from __future__ import annotations

import uuid
from datetime import UTC, date, datetime, timedelta
from decimal import Decimal

import factory
from faker import Faker

from app.models.micro_school import (
    MicroEnrollment,
    MicroEnrollmentStatus,
    MicroGroup,
    MicroPayment,
    MicroPaymentPeriodType,
    MicroPaymentStatus,
    MicroProgressLog,
    MicroResource,
    MicroResourceType,
    MicroSchool,
    MicroSchoolStatus,
)
from tests.factories.base import AsyncSQLAlchemyFactory
from tests.factories.iam import UserFactory

fake = Faker("fr_FR")


def _utc_now() -> datetime:
    return datetime.now(UTC)


def _moroccan_phone() -> str:
    return f"+2126{fake.msisdn()[3:11]}"


class MicroSchoolFactory(AsyncSQLAlchemyFactory):
    """Factory for micro-schools."""

    class Meta:
        model = MicroSchool
        exclude = ("educator",)

    id = factory.LazyFunction(uuid.uuid4)
    educator = factory.SubFactory(UserFactory)
    educator_id = factory.LazyAttribute(lambda o: o.educator.id)
    name = factory.LazyFunction(lambda: f"Micro-Ecole {fake.city()}")
    neighborhood = factory.LazyFunction(lambda: fake.street_name())
    city = factory.LazyFunction(lambda: fake.city())
    phone = factory.LazyFunction(_moroccan_phone)
    max_capacity = 20
    status = MicroSchoolStatus.ACTIVE.value


class MicroGroupFactory(AsyncSQLAlchemyFactory):
    """Factory for micro-school groups."""

    class Meta:
        model = MicroGroup
        exclude = ("micro_school",)

    id = factory.LazyFunction(uuid.uuid4)
    micro_school = factory.SubFactory(MicroSchoolFactory)
    micro_school_id = factory.LazyAttribute(lambda o: o.micro_school.id)
    name = factory.Sequence(lambda n: f"Groupe {n + 1}")
    age_range_min = 3
    age_range_max = 5


class MicroEnrollmentFactory(AsyncSQLAlchemyFactory):
    """Factory for micro-enrollments."""

    class Meta:
        model = MicroEnrollment
        exclude = ("micro_group", "parent")

    id = factory.LazyFunction(uuid.uuid4)
    micro_group = factory.SubFactory(MicroGroupFactory)
    parent = factory.SubFactory(
        UserFactory,
        school_id=factory.SelfAttribute(
            "..micro_group.micro_school.educator.school_id"
        ),
    )
    micro_group_id = factory.LazyAttribute(lambda o: o.micro_group.id)
    parent_id = factory.LazyAttribute(lambda o: o.parent.id)
    child_name = factory.LazyFunction(fake.first_name)
    date_of_birth = factory.LazyFunction(lambda: date.today() - timedelta(days=4 * 365))
    enrolled_at = factory.LazyFunction(_utc_now)
    status = MicroEnrollmentStatus.ACTIVE.value


class MicroPaymentFactory(AsyncSQLAlchemyFactory):
    """Factory for micro-payments."""

    class Meta:
        model = MicroPayment
        exclude = ("micro_school", "parent", "child_enrollment")

    id = factory.LazyFunction(uuid.uuid4)
    child_enrollment = factory.SubFactory(MicroEnrollmentFactory)
    micro_school = factory.LazyAttribute(
        lambda o: o.child_enrollment.micro_group.micro_school
    )
    parent = factory.LazyAttribute(lambda o: o.child_enrollment.parent)
    micro_school_id = factory.LazyAttribute(lambda o: o.micro_school.id)
    parent_id = factory.LazyAttribute(lambda o: o.parent.id)
    child_enrollment_id = factory.LazyAttribute(lambda o: o.child_enrollment.id)
    amount = Decimal("450.00")
    currency = "MAD"
    period_type = MicroPaymentPeriodType.MONTHLY.value
    period_start = factory.LazyFunction(lambda: date.today().replace(day=1))
    period_end = factory.LazyAttribute(lambda o: o.period_start + timedelta(days=29))
    paid_at = None
    status = MicroPaymentStatus.PENDING.value


class MicroResourceFactory(AsyncSQLAlchemyFactory):
    """Factory for micro learning resources."""

    class Meta:
        model = MicroResource

    id = factory.LazyFunction(uuid.uuid4)
    title = factory.LazyFunction(lambda: f"Ressource {fake.word().title()}")
    description = factory.LazyFunction(lambda: fake.sentence(nb_words=8))
    resource_type = MicroResourceType.ACTIVITY_SHEET.value
    age_group = "3-5"
    language = "fr"
    file_url = factory.LazyFunction(
        lambda: f"https://cdn.ecole.ma/{uuid.uuid4().hex}.pdf"
    )
    is_premium = False


class MicroProgressLogFactory(AsyncSQLAlchemyFactory):
    """Factory for micro progress logs."""

    class Meta:
        model = MicroProgressLog
        exclude = ("micro_enrollment", "educator")

    id = factory.LazyFunction(uuid.uuid4)
    micro_enrollment = factory.SubFactory(MicroEnrollmentFactory)
    educator = factory.LazyAttribute(
        lambda o: o.micro_enrollment.micro_group.micro_school.educator
    )
    micro_enrollment_id = factory.LazyAttribute(lambda o: o.micro_enrollment.id)
    educator_id = factory.LazyAttribute(lambda o: o.educator.id)
    date = factory.LazyFunction(date.today)
    note = factory.LazyFunction(lambda: "Participation active aux activites du jour.")
    photo_url = factory.LazyFunction(
        lambda: f"https://cdn.ecole.ma/{uuid.uuid4().hex}.jpg"
    )
    milestone_tag = "language"


__all__ = [
    "MicroSchoolFactory",
    "MicroGroupFactory",
    "MicroEnrollmentFactory",
    "MicroPaymentFactory",
    "MicroResourceFactory",
    "MicroProgressLogFactory",
]

File: mobile/lib/domain/entities/micro_school.dart
class MicroSchool {
  final String id;
  final String name;
  final String description;
  final String location;
  final String city;
  final int capacity;
  final int studentCount;
  final String status;

  const MicroSchool({
    required this.id,
    required this.name,
    required this.description,
    required this.location,
    required this.city,
    required this.capacity,
    required this.studentCount,
    required this.status,
  });

  double get capacityRatio => capacity == 0 ? 0 : studentCount / capacity;

  factory MicroSchool.fromJson(Map<String, dynamic> json) {
    return MicroSchool(
      id: json['id']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      description: json['description']?.toString() ?? '',
      location: json['location']?.toString() ?? '',
      city: json['city']?.toString() ?? '',
      capacity: (json['capacity'] as num?)?.toInt() ?? 0,
      studentCount: (json['student_count'] as num?)?.toInt() ?? 0,
      status: json['status']?.toString() ?? 'active',
    );
  }
}

class MicroEnrollment {
  final String id;
  final String microSchoolId;
  final String childName;
  final String status;

  const MicroEnrollment({
    required this.id,
    required this.microSchoolId,
    required this.childName,
    required this.status,
  });

  factory MicroEnrollment.fromJson(Map<String, dynamic> json) {
    return MicroEnrollment(
      id: json['id']?.toString() ?? '',
      microSchoolId: json['micro_school_id']?.toString() ?? '',
      childName: json['child_name']?.toString() ?? '',
      status: json['status']?.toString() ?? 'active',
    );
  }
}

class MicroPayment {
  final String id;
  final String microSchoolId;
  final double amount;
  final String currency;
  final String status;

  const MicroPayment({
    required this.id,
    required this.microSchoolId,
    required this.amount,
    required this.currency,
    required this.status,
  });

  factory MicroPayment.fromJson(Map<String, dynamic> json) {
    return MicroPayment(
      id: json['id']?.toString() ?? '',
      microSchoolId: json['micro_school_id']?.toString() ?? '',
      amount: (json['amount'] as num?)?.toDouble() ?? 0,
      currency: json['currency']?.toString() ?? 'MAD',
      status: json['status']?.toString() ?? 'pending',
    );
  }
}

class MicroResource {
  final String id;
  final String microSchoolId;
  final String title;
  final String description;
  final String resourceType;
  final String language;

  const MicroResource({
    required this.id,
    required this.microSchoolId,
    required this.title,
    required this.description,
    required this.resourceType,
    required this.language,
  });

  factory MicroResource.fromJson(Map<String, dynamic> json) {
    return MicroResource(
      id: json['id']?.toString() ?? '',
      microSchoolId: json['micro_school_id']?.toString() ?? '',
      title: json['title']?.toString() ?? '',
      description: json['description']?.toString() ?? '',
      resourceType: json['resource_type']?.toString() ?? '',
      language: json['language']?.toString() ?? 'fr',
    );
  }
}

class MicroMetricPoint {
  final String label;
  final double value;

  const MicroMetricPoint({
    required this.label,
    required this.value,
  });

  factory MicroMetricPoint.fromJson(Map<String, dynamic> json) {
    return MicroMetricPoint(
      label: json['label']?.toString() ?? '',
      value: (json['value'] as num?)?.toDouble() ?? 0,
    );
  }
}

class MicroProgressOverview {
  final double averageProgress;
  final int activeStudents;
  final double completionRate;
  final List<MicroMetricPoint> series;

  const MicroProgressOverview({
    required this.averageProgress,
    required this.activeStudents,
    required this.completionRate,
    required this.series,
  });

  factory MicroProgressOverview.fromJson(Map<String, dynamic> json) {
    final seriesJson = (json['series'] as List<dynamic>? ?? const []);
    return MicroProgressOverview(
      averageProgress: (json['average_progress'] as num?)?.toDouble() ?? 0,
      activeStudents: (json['active_students'] as num?)?.toInt() ?? 0,
      completionRate: (json['completion_rate'] as num?)?.toDouble() ?? 0,
      series: seriesJson
          .map((item) => MicroMetricPoint.fromJson(item as Map<String, dynamic>))
          .toList(),
    );
  }
}

class MicroStudentProgress {
  final String studentId;
  final String studentName;
  final int milestonesCompleted;
  final double progressRate;
  final List<MicroMetricPoint> series;

  const MicroStudentProgress({
    required this.studentId,
    required this.studentName,
    required this.milestonesCompleted,
    required this.progressRate,
    required this.series,
  });

  factory MicroStudentProgress.fromJson(Map<String, dynamic> json) {
    final seriesJson = (json['series'] as List<dynamic>? ?? const []);
    return MicroStudentProgress(
      studentId: json['student_id']?.toString() ?? '',
      studentName: json['student_name']?.toString() ?? '',
      milestonesCompleted: (json['milestones_completed'] as num?)?.toInt() ?? 0,
      progressRate: (json['progress_rate'] as num?)?.toDouble() ?? 0,
      series: seriesJson
          .map((item) => MicroMetricPoint.fromJson(item as Map<String, dynamic>))
          .toList(),
    );
  }
}

class MicroSchoolDetailBundle {
  final MicroSchool school;
  final List<MicroEnrollment> enrollments;
  final List<MicroResource> resources;
  final List<MicroPayment> payments;
  final MicroProgressOverview progress;

  const MicroSchoolDetailBundle({
    required this.school,
    required this.enrollments,
    required this.resources,
    required this.payments,
    required this.progress,
  });
}
